### PR TITLE
Implement custom data loader

### PR DESCRIPTION
**Pull request by shashitnak**
**Summary:**  
Testing out Dataloader changes from a fork of `async-graphql`

**Issue Reference(s):**  
Related to #419 

**Build & Testing:**
- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -57,6 +57,17 @@ num_cpus = "1.16.0"
...
+futures-channel = { version = "0.3.13" }
+futures-timer = { version = "3.0.2" }
+futures-util = { features = [
+  "std",
+  "io",
+  "sink",
+  "async-await",
+  "async-await-macro",
+] }
+lru = { version = "0.12.1" }
```
Comment (tusharmath) [inline Cargo.toml:70]: Really? We need all this?

File: src/http/graphql_data_loader.rs [hunk 2]
```diff
@@ -4,10 +4,10 @@ use std::sync::Arc;
...
 
 use async_graphql::async_trait;
-use async_graphql::dataloader::{DataLoader, Loader, NoCache};
 use async_graphql::futures_util::future::join_all;
 
 use crate::config::Batch;
+use crate::dataloader::{DataLoader, Loader, NoCache};
 use crate::http::{DataLoaderRequest, HttpClient, Response};
 
 pub struct GraphqlDataLoader {
```
Comment (tusharmath) [inline src/http/graphql_data_loader.rs:13]: This file should be in `/graphql` folder. Can do it here or in a different PR.

File: Cargo.toml (hunk 1 above)
Comment (shashitnak) [reply Cargo.toml:70]: we do need `lru` in cache.rs but I'm not sure about those async features will try to compile after removing those

Comment (shashitnak) [reply Cargo.toml:70]: I copied the cache code too incase we need it in the future

File: src/dataloader/mod.rs [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+use std::hash::Hash;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+pub use cache::{CacheFactory, CacheStorage, HashMapCache, LruCache, NoCache};
+use fnv::FnvHashMap;
+use futures_channel::oneshot;
+use futures_timer::Delay;
+use futures_util::future::BoxFuture;
```
Comment (tusharmath) [inline src/dataloader/mod.rs:15]: Split `mod.rs` into smaller files and keep a light weight mod instead.

File: src/http/graphql_data_loader.rs (hunk 2 above)
Comment (shashitnak) [reply src/http/graphql_data_loader.rs:13]: should be a small change. will do that in this PR itself

File: src/dataloader/cache.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+use std::collections::HashMap;
+use std::hash::{BuildHasher, Hash};
+use std::marker::PhantomData;
+use std::num::NonZeroUsize;
+
+/// Factory for creating cache storage.
+pub trait CacheFactory: Send + Sync + 'static {
+  /// Create a cache storage.
+  ///
+  /// TODO: When GAT is stable, this memory allocation can be optimized away.
```
Comment (meskill) [inline src/dataloader/cache.rs:12]: well, there is another point here to possible improvement - GAT is stable for a long time https://blog.rust-lang.org/2022/10/28/gats-stabilization.html

File: src/dataloader/mod.rs [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+    let mut requests = self.inner.requests.lock().unwrap();
+    let typed_requests = requests.get_mut(&tid).unwrap();
+    typed_requests.disable_cache = !enable;
+  }
+
+  /// Use this `DataLoader` load a data.
+  #[cfg_attr(feature = "tracing", instrument(skip_all))]
+  pub async fn load_one(&self, key: K) -> Result<Option<T::Value>, T::Error>
+  where
+    K: Send + Sync + Hash + Eq + Clone + 'static,
```
Comment (meskill) [inline src/dataloader/mod.rs:216]: instead of copying the same type conditions all over again we can use type alias

File: src/dataloader/mod.rs [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+    //         let loader = loader.clone();
+    //         move |n| {
+    //             let loader = loader.clone();
+    //             async move { loader.load_one(n).await }
+    //         }
+    //     }))
+    //     .await
+    //     .unwrap(),
+    //     (0..100).map(Option::Some).collect::<Vec<_>>()
+    // );
```
[the remark this answers is not in the document]
Comment (meskill) [reply src/dataloader/mod.rs:455]: the first part of test is totally valid. It'll be better to split this test in two or remove that part since is not supported by our fork

File: src/dataloader/mod.rs [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+
+  /// Type of error.
+  type Error: Send + Clone + 'static;
+
+  /// Load the data set specified by the `keys`.
+  async fn load(&self, keys: &[K]) -> Result<HashMap<K, Self::Value>, Self::Error>;
+}
+
+struct DataLoaderInner<K: Send + Sync + Hash + Eq + Clone + 'static, T: Loader<K>> {
+  requests: Mutex<FnvHashMap<TypeId, Requests<K, T>>>,
```
Comment (meskill) [inline src/dataloader/mod.rs:65]: will anyway leave a comment here just as a reminder

in next iteration we should try to remove wrapper `FnvHashMap` over Requests since there is not possible anymore to have different types for the same loader

Comment (shashitnak) [reply src/dataloader/mod.rs]: made this change and it's working now

File: src/dataloader/cache.rs (hunk 4 above)
Comment (shashitnak) [reply src/dataloader/cache.rs]: I am not sure if I used this GAT thing properly but I did remove that `dyn` and made it statically type
